e2e/nomad09upgrade/upgrade.go

```diff
@@ -0,0 +1,339 @@
+//+build linux
+
+package nomad09upgrade
+
+import (
+	"bytes"
+	"fmt"
+	"html/template"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	getter "github.com/hashicorp/go-getter"
+	"github.com/hashicorp/nomad/api"
+	"github.com/hashicorp/nomad/e2e/e2eutil"
+	"github.com/hashicorp/nomad/e2e/framework"
+	"github.com/hashicorp/nomad/helper/discover"
+	"github.com/hashicorp/nomad/helper/testlog"
+	"github.com/hashicorp/nomad/helper/uuid"
+	"github.com/hashicorp/nomad/testutil"
+	"github.com/stretchr/testify/require"
+)
+
+func init() {
+	framework.AddSuites(&framework.TestSuite{
+		Component:   "nomad09upgrade",
+		CanRunLocal: true,
+		Cases: []framework.TestCase{
+			&UpgradePathTC{},
+		},
+	})
+}
+
+var (
+	nomadVersions = []string{
+		"0.8.7",
+		"0.8.6",
+		"0.8.5",
+		"0.8.4",
+		"0.8.3",
+		"0.8.2",
+		"0.8.1",
+		"0.8.0",
+	}
+
+	agentTemplate = `
+ports {
+  http = {{.HTTP}}
+  rpc  = {{.RPC}}
+  serf = {{.Serf}}
+}
+
+server {
+	enabled = true
+	bootstrap_expect = 1
+}
+
+client {
+	enabled = true
+	options = {
+      "driver.raw_exec.enable" = "1"
+    }
+}
+`
+)
+
+type templateVars struct {
+	HTTP int
+	RPC  int
+	Serf int
+}
+
+func writeConfig(path string, vars templateVars) error {
+	t := template.Must(template.New("config").Parse(agentTemplate))
+	f, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	return t.Execute(f, vars)
+}
+
+type UpgradePathTC struct {
+	framework.TC
+
+	binDir string
+	bin    string
+}
+
+type nomadAgent struct {
+	bin     string
+	dataDir string
+	stdout  bytes.Buffer
+	stderr  bytes.Buffer
+	cmd     *exec.Cmd
+	vars    templateVars
+	killed  bool
+
+	targetCmd *exec.Cmd
+}
+
+func (tc *UpgradePathTC) newNomadServer(t *testing.T, ver string) (*nomadAgent, error) {
+	dir, err := ioutil.TempDir("/opt", "nomade2e")
+	if err != nil {
+		return nil, err
+	}
+
+	return tc.newNomadServerWithDataDir(t, ver, dir)
+}
+
+func (tc *UpgradePathTC) newNomadServerWithDataDir(t *testing.T, ver string, dataDir string) (*nomadAgent, error) {
+	srv := &nomadAgent{
+		bin:     filepath.Join(tc.binDir, ver, "nomad"),
+		dataDir: dataDir,
+	}
+	conf := filepath.Join(dataDir, "config.hcl")
+	httpPort, err := getFreePort()
+	if err != nil {
+		return nil, err
+	}
+	rpcPort, err := getFreePort()
+	if err != nil {
+		return nil, err
+	}
+	serfPort, err := getFreePort()
+	if err != nil {
+		return nil, err
+	}
+
+	srv.vars = templateVars{
+		HTTP: httpPort,
+		RPC:  rpcPort,
+		Serf: serfPort,
+	}
+	if err := writeConfig(conf, srv.vars); err != nil {
+		return nil, err
+	}
+
+	w := testlog.NewWriter(t)
+
+	srv.cmd = exec.Command(srv.bin, "agent", "-config", conf, "-log-level", "DEBUG", "-data-dir", srv.dataDir)
+	srv.cmd.Stdout = w
+	srv.cmd.Stderr = w
+
+	srv.targetCmd = exec.Command(tc.bin, "agent", "-config", conf, "-log-level", "TRACE", "-data-dir", srv.dataDir)
+	srv.targetCmd.Stdout = w
```

Comment:
You could wrap the log writer in a prefix to separate primary vs target agent log lines: https://golang.org/pkg/testing/iotest/#NewWriteLogger